src/utils/PathUtils.ts

```diff
@@ -0,0 +1,132 @@
+      }
+      brackets.forEach(bracket => {
+        if (!indexRegex.test(bracket) && !(bracket === '+' || bracket === '=')) {
+          seenSlices.push(bracket);
+        }
+      });
+      if (seenSlices.length > 0) {
+        pathParts.push({ base: fhirPathBase, brackets: brackets, slices: seenSlices });
+      } else {
+        pathParts.push({ base: fhirPathBase, brackets: brackets });
+      }
+    }
+  }
+  return pathParts;
+}
+
+/**
+ * Assembles a PathPart array back to its original form
+ * @param {PathPart[]} pathParts - An array of pathParts
+ * @returns {string} path - The path corresponding to those pathParts
+ */
+export function assembleFSHPath(pathParts: PathPart[]): string {
+  let path = '';
+  pathParts.forEach((pathPart, i) => {
+    path += pathPart.base;
+    pathPart.brackets?.forEach(bracket => (path += `[${bracket}]`));
+    if (i < pathParts.length - 1) {
+      path += '.';
+    }
+  });
+  return path;
+}
+
+/**
+ *
+ * @param {PathPart} element - A single element in a rules path
+ * @param {Map<string, number} pathMap - A map containing an element's name as the key and that element's updated index as the value
+ */
+function convertSoftIndexes(element: PathPart, pathMap: Map<string, number>) {
+  const mapName = element.slices ? element.base.concat(element.slices.join('|')) : element.base;
+  if (!pathMap.has(mapName)) {
+    pathMap.set(mapName, 0);
+    if (element.brackets?.includes('+')) {
+      element.brackets[element.brackets.indexOf('+')] = '0';
+    }
+  } else {
+    element.brackets?.forEach((bracket: string, index: number) => {
+      if (bracket === '+') {
+        const newIndex = pathMap.get(mapName) + 1;
+        element.brackets[index] = newIndex.toString();
+        pathMap.set(mapName, newIndex);
+      } else if (bracket === '=') {
+        const currentIndex = pathMap.get(mapName);
+        element.brackets[index] = currentIndex.toString();
+      }
+    });
+  }
+}
+
+/**
+ * Replaces soft indexs in rule paths with corresponding numbers
+ * @param {Rule[]} rules - An array of Rules
+ */
+export function resolveSoftIndexing(rules: Array<Rule | CaretValueRule>): void {
+  const pathMap: Map<string, number> = new Map();
+  const caretPathMap: Map<string, Map<string, number>> = new Map();
+
+  // Parsing and separating rules by base name and bracket indexes
+  const parsedRules = rules.map(rule => {
+    const parsedPath: { path: PathPart[]; caretPath?: PathPart[] } = {
+      path: parseFSHPath(rule.path)
+    };
+    // If we have a CaretValueRule, we'll need a second round of parsing for the caret path
+    if (rule instanceof CaretValueRule) {
+      parsedPath.caretPath = parseFSHPath(rule.caretPath);
+    }
+    return parsedPath;
+  });
+
+  // Replacing Soft indexes with numbers
+  parsedRules.forEach((parsedRule, index) => {
+    const originalRule = rules[index];
+    parsedRule.path.forEach((element: PathPart) => {
+      convertSoftIndexes(element, pathMap);
+    });
```

Comment:
I don't think this approach is quite right, because it's calling `convertSoftIndexes` with only the `PathPart` (any portion of the path between `.`).  So, for example, for the path `foo[+].bar[+]`, `convertSoftIndexes` will first get called w/ `foo[+]` and then `bar[+]` -- and each will be the key in the shared map.  In this example, it happens to work, but consider a path that has nested items w/ the same name: `item[+].item[+]`.  That should convert to `item[0].item[0]` (since `item` and `item[0].item` are distinct arrays), but instead gets converted to `item[0].item[1]`.

I think the map needs the prefixes of the pathParts too (and the prefixes need to be already resolved).  So in my first example case, it should be tracking `foo` and `foo[0].bar` in the map.

Reply:
You're totally right. This is something that occurred to me as a potential edge case, but I wasn't sure if recursive paths where allowed in the FHIR spec. I'll update this function to support edge cases such as these. Also, I think I separate `PathUtils` test file makes a lot of sense!